server/controllers/finance/purchases.js

```diff
@@ -0,0 +1,255 @@
+const BadRequest = require('./../../lib/errors/BadRequest');
+
+// create a new purchase order
+exports.create = create;
+
+// read all purchase order
+exports.list = list;
+
+// read a specific purchase order
+exports.detail = detail;
+
+// Update properties of a purchase Order
+exports.update = update;
+
+
+/**
+ * @function linkPurchaseItems
+ *
+ * @description
+ * Utility method to ensure purchase item lines reference the purchase
+ * order and escape values as necessary
+ *
+ * @param {Array} items - an array of all purchase items to be written
+ * @param {Blob} purchaseUuid - UUID of referenced purchase order
+ * @returns {Array} - an array of all purchases items properly formatted
+ */
+function linkPurchaseItems(items, purchaseUuid) {
+
+  // this columns array exists so that we are sure to order to columns in the
+  // correct order
+  const columns = [
+    'uuid', 'inventory_uuid', 'quantity', 'unit_price', 'total', 'purchase_uuid'
+  ];
+
+  // loop through each item, making sure we have escapes and orderings correct
+  return items.map(item => {
+
+    // make sure that each item has a uuid by generate
+    item.uuid = db.bid(item.uuid || uuid.v4());
+    item.purchase_uuid = purchaseUuid;
+    item.inventory_uuid = db.bid(item.inventory_uuid);
+
+    // collapse sale items into array to be inserted into database
+    return columns.map(key => item[key]);
+  });
+}
+
+/**
+ * @function lookupPurchaseOrder
+ *
+ * @description
+ * Looks up a single purchase record and associated purchase_items
+ *
+ * @param {string} uid - the  UUID to be fetched from the database.
+ * @return {Promise} - the database's promise result
+ *
+ * @private
+ */
+function lookupPurchaseOrder(uid) {
+  let record;
+
+  let sql = `
+    SELECT BUID(p.uuid) AS uuid, CONCAT(pr.abbr, p.reference) AS reference,
+      p.cost, p.date, c.text AS supplier, p.user_id,
+      BUID(p.creditor_uuid) as creditor_uuid, p.note
+    FROM purchase AS p
+    JOIN creditor AS c ON c.uuid = p.creditor_uuid
+    JOIN project AS pr ON p.project_id = pr.id
+    WHERE p.uuid = ?;
+  `;
+
+  return db.exec(sql, [db.bid(uid)])
+  .then(function (rows) {
+    if (!rows.length) {
+      throw new NotFound(`Could not find a purchase with uuid ${uid}`);
+    }
+
+    // store the record for return
+    record = rows[0];
+
+    sql = `
+      SELECT BUID(pi.uuid) AS uuid, pi.quantity, pi.unit_price, pi.total, i.text
+      FROM purchase_item AS pi
+      JOIN inventory AS i ON i.uuid = pi.inventory_uuid
+      WHERE pi.purchase_uuid = ?;
+    `;
+
+    return db.exec(sql, [db.bid(uid)]);
+  })
+  .then(function (rows) {
+
+    // bind the purchase items to the "items" property and return
+    record.items = rows;
+    return record;
+  });
+}
+
+/**
+ * @method create
+ *
+ * @description
+ * POST /purchases
+ *
+ * Creates a purchase order in the database
+ */
+function create(req, res, next) {
+  let data = req.body;
+
+  if (!data.items) {
+    return next(
```

Comment:
I've noticed that in some places you are throwing Error objects and in others passing them through to `next`.

In methods that do not have access to `next` we would have to use throw but it seems to be that `throw` here would pass the object through to `next` in the same way.

Do you have a preference?